File: 알고리즘 스터디/개인공부/Greedy/TimeMachine.py
# https://www.acmicpc.net/problem/11657
# 밸만포드 알고리즘
import sys

if __name__ == '__main__':
    n_node, n_edge = map(int, sys.stdin.readline().split())
    cost = [5000001] * n_node
    edge = [dict() for _ in range(n_node)]
    inf_loop = False

    for _ in range(n_edge):
        src, dst, _cost = map(int, sys.stdin.readline().split())
        edge[src - 1][dst - 1] = min(_cost, edge[src - 1].get(dst - 1, 5000001))

    # 벨만포드
    cost[0] = 0     # 출발지 -> 출발지 비용 : 0
    for _ in range(n_node - 1):
        for cur_node in range(n_node):
            if cost[cur_node] != 5000001:
                # 가중치 갱신
                for neigh in edge[cur_node].keys():
                    cost[neigh] = min(cost[neigh], cost[cur_node] + edge[cur_node][neigh])
    
    # 여기서 갱신이 일어나면 무한 음수 간선 존재하는 거임
    for cur_node in range(n_node):
        if cost[cur_node] != 5000001:
            # 가중치 갱신
            for neigh in edge[cur_node].keys():
                if cost[neigh] > cost[cur_node] + edge[cur_node][neigh]:
                    inf_loop = True
                    break

            if inf_loop:
                break

    if inf_loop:
        print(-1)
    else:
        # 비용 출력
        for i in range(1, n_node):
            if cost[i] == 5000001:
                print(-1)
            else:
                print(cost[i])

File: 알고리즘 스터디/개인공부/Greedy/TimeMachine.cpp
#include<iostream>
#include<vector>
#include<map>

using namespace std;

int main() {
	long long n_node, n_edge;
	vector<long long> costs;
	vector<map<long long, long long>> edge;
	bool inf_loop = false;

	cin >> n_node >> n_edge;
	costs = vector<long long>(n_node, 5000001);
	edge = vector<map<long long, long long>>(n_node);

	for (long long i = 0, src, dst, cost; i < n_edge; i++) {
		scanf("%lld %lld %lld", &src, &dst, &cost);
		
		if (edge[src - 1].find(dst - 1) == edge[src - 1].end())
			edge[src - 1][dst - 1] = 5000001;

		edge[src - 1][dst - 1] = min(edge[src - 1][dst - 1], cost);
	}

	// 벨만포드
	costs[0] = 0;	// 출발지 -> 출발지 : 비용 0
	for (long long i = 0; i < n_node - 1; i++) {
		for (long long cur_node = 0; cur_node < n_node; cur_node++) {
			if (costs[cur_node] != 5000001) {
				// 가중치 갱신
				for (map<long long, long long>::iterator it = edge[cur_node].begin(); it != edge[cur_node].end(); it++)
					costs[it->first] = min(costs[it->first], costs[cur_node] + it->second);
			}
		}
	}

	// 여기서 갱신이 일어나면 무한 음수 간선이 존재하는 거임
	for (long long cur_node = 0; cur_node < n_node; cur_node++) {
		if (costs[cur_node] != 5000001) {
			// 가중치 갱신
			for (map<long long, long long>::iterator it = edge[cur_node].begin(); it != edge[cur_node].end(); it++) {
				if (costs[it->first] > costs[cur_node] + it->second) {
					inf_loop = true;
					goto collapse;
				}
			}
		}
	}

collapse:
	if (inf_loop)
		cout << -1 << "\n";
	else {
		// 비용 출력
		for (int i = 1; i < n_node; i++) {
			if (costs[i] == 5000001)
				printf("-1\n");
			else
				printf("%lld\n", costs[i]);
		}
	}

	return 0;
}